### ISSUE TITLE
Parallel packaging

### ISSUE THREAD

**Issue by fivepapertigers** [serverless/serverless#5002]
# This is a Feature Proposal(?)

## Description

For feature proposals:
* What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.

This is somewhat related to #4485. I have multiple serverless services in a single codebase, each using a different child template to deploy a separate set of resources. When I package and deploy each sequentially, everything works as intended - great! However, I ran into an issue when trying to write a simple bash script that executes `sls package` in parallel. I'm using the `--package` flag to create separate deployment packages for each service, and my assumption was that this would protect collisions in parallel execution. However, it seems that even though I specify `--package`, Serverless first creates the `.zip` artifacts files inside a common `./.serverless` directory, _then_ moves them into their respective paths. When each process is done, it cleans up after itself by wiping out the `./.serverless` directory entirely. This leads to a race condition - one process clears the `./.serverless` directory, including another process's `.zip` artifact file.

I didn't mark this is as a bug, because it's not the expected use-case. But on the surface, it doesn't seem terribly difficult to build the artifacts in the location specified by `--package` instead of building them in `.serverless` and _then_ moving them.

* If there is additional config how would it look

N/A

Similar or dependent issues:

Couldn't find any

## Additional Data

* ***Serverless Framework Version you're using***:

1.25.0

* ***Operating System***:

OS X 10.12.6

* ***Stack Trace***:
* ***Provider Error messages***:
```

  Error --------------------------------------------------

  ENOENT: no such file or directory, open '<path-to-service>/.serverless/<service-name>.zip'
```

**Comment by Graham42** [comment]
I've been exploring a fix for this, I see there's been a couple previous attempts
- https://github.com/serverless/serverless/pull/5413
- https://github.com/serverless/serverless/pull/5344

I'm part way into a possible fix and just started reading the contributing guide and realized I should mention my approach incase there's some early feedback.

I'm planning on abstracting all direct references to the `.serverless` directory. For the references that don't use the `package` option, I'm keeping a separate function that will default to the existing behaviour for backwards compatibility. But, if there's another CLI parameter, maybe `--usePackageForTmp`, then all the references which previously hardcoded the `.serverless` directory, will use the `package` option.

**Comment by petergaultney** [comment]
My only comment is that I don't think the usefulness of abstracting this directory is necessarily limited to packaging. The ability to specify a general purpose working directory that lives outside the current directory is extremely valuable in cases with read-only filesystems, for instance. So abstracting away _all_ references to `.serverless` would be a general purpose improvement to the framework and probably shouldn't be limited to cases where a packaging-related option has been specified.

**Comment by bvalosek** [comment]
Had a very similar issue and race condition, but with a write conflict. I'd like to comment for anybody who ends up going down a similar rabbit hole that i did:

We run `sls deploy` in parallel via lerna's `lerna run` command. We had two services that both required the AWS custom resources `custom-resource.zip` file to be generated (in the common build directory) and then moved to the correct service directory. This created a race condition wherein the two `deploy` commands would both assume the cached file wasn't there, and proceeded to create it, with all but one erroring out due to the file already existing in the common directory with the following:

```
@yonomi/service-debugger: Serverless: Installing dependencies for custom CloudFormation resources...
@yonomi/service-permissions: Serverless: Packaging service...
@yonomi/service-permissions: Serverless: Installing dependencies for custom CloudFormation resources...
@yonomi/service-debugger: Serverless: Using deployment bucket 'yp-deploy-bucket-232674606574'
@yonomi/service-permissions:  
@yonomi/service-permissions:   Error --------------------------------------------------
@yonomi/service-permissions:  
@yonomi/service-permissions:   Error: dest already exists.
@yonomi/service-permissions:       at /home/circleci/yonomi-platform/node_modules/serverless/node_modules/fs-extra/lib/move/move.js:41:31
@yonomi/service-permissions:       at /home/circleci/yonomi-platform/node_modules/serverless/node_modules/universalify/index.js:21:54
@yonomi/service-permissions:  
@yonomi/service-permissions:      For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
```

**Comment by NoxHarmonium** [comment]
I hope I'm not derailing the original issue, but I have the exact same intermittent crash as the last poster when packaging/deploying two or more serverless projects in parallel.

```
➤ YN0000: [xxxx]: Serverless: Installing dependencies for custom CloudFormation resources...
➤ YN0000: [yyyy]: Serverless: Installing dependencies for custom CloudFormation resources...
➤ YN0000: [xxxx]:  
➤ YN0000: [xxxx]:  Error ---------------------------------------------------
➤ YN0000: [xxxx]:  
➤ YN0000: [xxxx]:  
➤ YN0000: [xxxx]:   Error: dest already exists.
➤ YN0000: [xxxx]:       at /var/lib/jenkins/workspace/xxxx-build_develop/node_modules/fs-extra/lib/move/move.js:41:31
➤ YN0000: [xxxx]:       at /var/lib/jenkins/workspace/xxxx-build_develop/node_modules/fs-extra/node_modules/universalify/index.js:22:54
➤ YN0000: [xxxx]:  
➤ YN0000: [xxxx]:      For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
```

The line that seems to be crashing is https://github.com/serverless/serverless/blob/ae67bf448950b199eaf2a0ec406349d63f60424c/lib/plugins/aws/customResources/generateZip.js#L21

I'm pretty sure the reason is that the ensureArtifact function is not thread safe. The custom CloudFormation resources are built once and cached in a shared folder in the home drive:

https://github.com/serverless/serverless/blob/ae67bf448950b199eaf2a0ec406349d63f60424c/lib/utils/ensureArtifact.js#L9

It seems that when there are multiple serverless instances building the artifacts and trying to cache them in the shared folder the crash happens.

I'd say that the easiest fix would be just to disable the caching because it doesn't even take that long to generate the custom resource package in the scheme of things. Other possible fixes I can think of are:

1. Make the cache directory specific to a project
2. Make the ensureArtifact method atomic/thread safe using a lock file or similar

I might create a new issue or find an existing one to move this comment to this issue isn't the right one
